### PR TITLE
[FLINK-30006][table-planner] Make all pure dynamic functions non-deterministic in streaming mode to avoid wrong plans

### PR DESCRIPTION
**Pull request by lincoln-lil**
## What is the purpose of the change
Make all pure dynamic functions non-deterministic in streaming mode to avoid wrong plans

## Brief change log
* change all pure dynamic functions non-deterministic in streaming mode

## Verifying this change
TODO new tests to be added

## Does this pull request potentially affect one of the following parts:
  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with @Public(Evolving): (no)
  - The serializers: (no )
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation
  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by XComp** [comment]
just as a hint after seeing your `(do not merge)`: Github supports PR drafts (see [this blog post](https://github.blog/2019-02-14-introducing-draft-pull-requests/) for instance). I hope that helps. :-)

**Comment by lincoln-lil** [comment]
@XComp  thank you very much for the warm tips！I'll try it later  : )

**Review by godfreyhe — APPROVED** [review]
LGTM

File: docs/content/docs/dev/table/functions/udfs.md [hunk 1]
```diff
@@ -647,15 +647,15 @@ The following system functions are always non-deterministic(evaluated per record
 - RAND_INTEGER
 - CURRENT_DATABASE
 - UNIX_TIMESTAMP
+- CURRENT_TIMESTAMP
 - CURRENT_ROW_TIMESTAMP
+- LOCALTIMESTAMP
+- NOW
```
Comment (godfreyhe) [inline docs/content/docs/dev/table/functions/udfs.md:653]: These methods are dynamic function, see CurrentTimePointCallGen

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/DirectConvertRule.java [hunk 2]
```diff
@@ -39,220 +41,281 @@
...
+                                definitionSqlOperatorHashMap.put(
+                                        BuiltInFunctionDefinitions.CURRENT_DATE, func);
+                            } else if (func.getName()
+                                    .equalsIgnoreCase(SqlStdOperatorTable.CURRENT_TIME.getName())) {
+                                definitionSqlOperatorHashMap.put(
+                                        BuiltInFunctionDefinitions.CURRENT_TIME, func);
+                            } else if (func.getName()
+                                    .equalsIgnoreCase(SqlStdOperatorTable.LOCALTIME.getName())) {
+                                definitionSqlOperatorHashMap.put(
+                                        BuiltInFunctionDefinitions.LOCAL_TIME, func);
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/DirectConvertRule.java:75]: It's better all dynamic functions should be defined here, including CURRENT_TIMESTAMP, LOCALTIMESTAMP, NOW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java [hunk 3]
```diff
@@ -60,20 +63,52 @@
...
         return instance;
     }
 
+    public static List<SqlFunction> dynamicFunctions(boolean isBatchMode) {
+        return Arrays.asList(
+                new FlinkTimestampDynamicFunction(
+                        SqlStdOperatorTable.LOCALTIME.getName(), SqlTypeName.TIME, isBatchMode),
+                new FlinkTimestampDynamicFunction(
+                        SqlStdOperatorTable.CURRENT_TIME.getName(), SqlTypeName.TIME, isBatchMode),
+                new FlinkCurrentDateFunction(isBatchMode));
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java:95]: ditto

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/FunctionGenerator.scala [hunk 4]
```diff
@@ -544,6 +541,23 @@ class FunctionGenerator private (tableConfig: ReadableConfig) {
...
+  FlinkSqlOperatorTable
+    .dynamicFunctions(!isStreamingMode)
+    .forEach(
+      func => {
+        if (func.getName == SqlStdOperatorTable.LOCALTIME.getName) {
+          addSqlFunction(func, Seq(), new CurrentTimePointCallGen(true, isStreamingMode))
+        } else if (
+          func.getName == SqlStdOperatorTable.CURRENT_DATE.getName
+          || func.getName == SqlStdOperatorTable.CURRENT_TIME.getName
+        ) {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/FunctionGenerator.scala:553]: diito

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/common/DynamicFunctionPlanTestBase.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                        + "     gmt_date, ts, cat, SUBSTR(CAST(ts AS VARCHAR), 1, 2), SUM(cnt)\n"
+                        + "FROM src\n"
+                        + "WHERE gmt_date = CURRENT_DATE\n"
+                        + "  AND cat = 'fruit' AND ts = CURRENT_TIME\n"
+                        + "GROUP BY gmt_date, ts, cat");
+    }
+
+    @Test
+    public void testCalcMerge() {
+        // TODO after FLINK-30841 fixed, this plan should be updated
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/common/DynamicFunctionPlanTestBase.java:93]: FLINK-30841  is accepted

**Review by godfreyhe — CHANGES_REQUESTED** [review]
Thanks for the contribution, I left some comments

File: docs/content/docs/dev/table/functions/udfs.md (hunk 1 above)
Comment (lincoln-lil) [reply docs/content/docs/dev/table/functions/udfs.md:653]: yes, for the runtime implemantation, they all behavior as dynamic functions, I'll update this

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/DirectConvertRule.java (hunk 2 above)
Comment (lincoln-lil) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/DirectConvertRule.java:75]: This should be done and also for the FlinkSqlOperatorTable, easier for maintainance

**Review by lincoln-lil — COMMENTED** [review]
@godfreyhe thank you for your comments! I'll rebase and update this pr

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkTimestampDynamicFunction.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+ * Function that used to define SQL time functions like LOCALTIME, CURRENT_TIME(these are all
+ * dynamic functions in Calcite's {@link SqlStdOperatorTable}) in Flink, the difference from the
+ * parent {@link SqlAbstractTimeFunction} is this function class be aware of whether it is used in
+ * batch mode, if true it will act totally same as the parent {@link SqlAbstractTimeFunction}, but
+ * will be a non-deterministic function if not in batch mode.
+ */
+@Internal
+public class FlinkTimestampDynamicFunction extends SqlAbstractTimeFunction {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkTimestampDynamicFunction.java:38]: just simple named: FlinkTimestampFunction?
or we also should rename `FlinkCurrentDateFunction` to `FlinkCurrentDateDynamicFunction`

**Review by godfreyhe — COMMENTED** [review]
I have one minor comment

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkTimestampDynamicFunction.java (hunk 6 above)
Comment (lincoln-lil) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkTimestampDynamicFunction.java:38]: Agree rename the `FlinkCurrentDateFunction ` to `FlinkCurrentDateDynamicFunction`, this unification is necessary, I'll update it.

**Review by godfreyhe — APPROVED** [review]
LGTM

**Comment by lincoln-lil** [comment]
@flinkbot run azure

**Comment by lincoln-lil** [comment]
@flinkbot run azure

**Comment by lincoln-lil** [comment]
The e2e ci failed cause by unstable connection(FLINK-30921) in flink ci env(https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=45753&view=results), but in my private pipeline it succeed（ https://dev.azure.com/lincoln86xy/flink/_build/results?buildId=455&view=results ）

**Pull request merged**